### ISSUE TITLE
Local Tools Early Preview Documentation

### ISSUE THREAD

**Issue by wli3** [dotnet/cli#10288]
# Local Tools Early Preview Documentation

**This document is for checkin build of dotnet SDK. The feature is still work in progress. We hope to gather more feedback especially on user experience by this documentation**. If you think the experience need improvement, you can file issues at [Github](https://github.com/dotnet/cli/issues)

## Motivation

Enable high cohesion between tools and corresponding code repositories. The repository maintainer could check-in a list of dotnet tools with certain version. And other contributor can use these tools within the repository with certain gestures. Local tools use the same package as global tools.

### Install the latest dotnet SDK "checkin build"

1. You can find the installer at [here](https://github.com/dotnet/core-sdk#installers-and-binaries) under the _Master (3.0.x Runtime)_ column.
2. Most of the existing dotnet tools package(global tools) are targeting runtime netcoreapp2.1. You need to install it in order to run the existing command. You can find it at [here](https://www.microsoft.com/net/download/dotnet-core/2.1) under _Run apps - Runtime_ column.

### Create tool manifest file

The _create_ experience has not been implemented yet. However you can create a tool manifest file by copy paste from the following template. And name it `dotnet-tools.json`

``` json
{
    "isRoot": true,
    "tools": {
        "dotnetsay": {
            "version": "2.1.4",
            "commands": [
                "dotnetsay"
            ]
        }
    }
}
```

### Restore command

In order to use the command specified in tool manifest file. You need to run command `dotnet tool restore` and you should see the following success message:

```shell
Tool 'dotnetsay' (version '2.1.4') was restored. Available commands: dotnetsay


Restore was successful.
```

### Run the command

Run `dotnet dotnetsay hi`. You should see _dotentsay_ being executed.

If you see the following error, please install netcoreapp2.1 runtime. At https://www.microsoft.com/net/download/dotnet-core/2.1. Since _dotnetsay_ and most of the global tools are targeting 2.1 runtime while preview dotnet SDK only contains preview netcoreapp3.0 runtime.

```shell
It was not possible to find any compatible framework version
The specified framework 'Microsoft.NETCore.App', version '2.1.0' was not found.
  - Check application dependencies and target a framework version installed at:
      C:\Users\wul\Downloads\dotnet-sdk-latest-win-x64-localtools\
  - Installing .NET Core prerequisites might help resolve this problem:
      https://go.microsoft.com/fwlink/?LinkID=798306&clcid=0x409
  - The .NET Core framework and SDK can be installed from:
      https://aka.ms/dotnet-download
  - The following versions are installed:
      3.0.0-preview1-27017-04 at [C:\Users\wul\Downloads\dotnet-sdk-latest-win-x64-localtools\shared\Microsoft.NETCore.App]
```

### The _scope_ of the tool manifest file

If there is a tool manifest file or a tool manifest file under folder _.config_ in the parent/current directory of the current directory. The current directory has the access to the tools specified in the tool manifest file. For example. If there is a tool manifest with a tool _dotnetsay_ in the root of the repository, the user could run `dotnet dotnetsay hi` in any sub directory of the repository.

### isRoot flag

The flag in the manifest file means "stop looking up parent directories for manifest file when the first one is found". This entry in manifest file is required.

### Invoke local tools

Global tools use $PATH and shell to resolve command via shim. However we could not use the same mechanism due to the limitation of the operation system shell. As a result, local tools require prefix `dotnet`. The following is the map of commands.

#### Global tools to local tools command map

| global tools command invoke  | local tools command invoke  |
| ------ | ------ |
| dotnetsay  | dotnet dotnetsay  |
| dotnet say (aka dotnet-say)  | dotnet say  |
| dotnetsay and dotnet say both exist  | dotnet dotnetsay and dotnet dotnet-say |

### Example of multiple tools

```json
{
    "version": "1",
    "isRoot": true,
    "tools": {
        "t-rex": {
            "version": "1.0.53",
            "commands": ["t-rex"],
        },
        "dotnetsay": {
            "version": "2.1.4",
            "commands": ["dotnetsay"]
        }
    }
}
```

### Requirement of commands

We require listing out the command in the tool package. Note many dotnet tool packages do not have matching package id and the command. The feature of adding command to the tool manifest file has not been implemented yet. This step will be easier when it is completed.

### Term

Local tool scope: the working directory that has the access to the tool specified in the `dotnet-tools.json` stored in current or parent directory.

Tool manifest file: `dotnet-tools.json`

Repository maintainer: A dotnet tool consumer. Programmers have the write access to the repository and decide which dotnet tools should be used in this repository.

Repository contributor: A dotnet tool consumer. Programmers have the read access to the repository and use dotnet tools decided by repository maintainer.

Local tools restore: Restore all tools that is possible to be accessed by current directory and generate cache.

Resolver cache: Technical, not customer facing. Cache to speed up the local tool resolution.

**Comment by wli3** [comment]
I will close this issue after preview is done

**Comment by jmezach** [comment]
Am I correct in assuming that this would allow different versions of a tool within different repositories? For example, in the npm world we can have different versions of say typescript within different repositories. What happens if a global tool is also available? Does the local one basically override the globally installed one, provided that you'd run it by using dotnet <tool>?

**Comment by wli3** [comment]
> Am I correct in assuming that this would allow different versions of a tool within different repositories? For example, in the npm world we can have different versions of say typescript within different repositories.

Yes. That is one of the problem we want to solve.


> What happens if a global tool is also available? Does the local one basically override the globally installed one, provided that you'd run it by using dotnet ?

Run local tool would require prefix `dotnet`. So it will not be ambiguous most of the time. And when there is a global tools with the pattern "dotnet-*" CLI will find the tool with the following rule

| global tools command invoke  | local tools command invoke  |
| ------ | ------ |
| dotnetsay  | dotnet dotnetsay  |
| dotnet say (aka dotnet-say)  | dotnet say  |
| dotnetsay and dotnet say both exist  | dotnet dotnetsay and dotnet dotnet-say |

**Comment by wli3** [comment]
Update:
To invoke local tools you need to use `dotnet tool run dotnetsay hi`.

**Comment by wli3** [comment]
Added _Manage local tools_ session

**Comment by amis92** [comment]
After latest change (a move back from `dotnet tool run fulltoolname` to `dotnet fulltoolname`), shouldn't that:

> global tools command invoke | local tools command invoke
> -- | --
> dotnetsay | dotnet dotnetsay
> dotnet say (aka dotnet-say) | dotnet say
> dotnetsay and dotnet say both exist | dotnet dotnetsay and dotnet dotnet-say

 Be rather


global tools command invoke | local tools command invoke
-- | --
dotnetsay | dotnet dotnetsay
dotnet say (aka dotnet-say) | dotnet dotnet-say
dotnetsay and dotnet say both exist | dotnet dotnetsay and dotnet dotnet-say

For some reason this single case "infers" the dotnet part. I think it shouldn't as well?

**Comment by wli3** [comment]
@amis92 sorry that was a mistake in the document. We still use `dotnet tool run dotnetsay hi` today. And we are still discussing if we need a short hand version or not.

**Comment by Thraka** [comment]
What about compromising on that with `dotnet tool dotnetsay`? Basically, any command after `dotnet tool` that isn't recognized is parsed as if `dotnet tool run` was used.

**Comment by kzu** [comment]
having local + global automatic fallbacks with shorthand names would be awesome.

**Comment by wli3** [comment]
bring back `dotnet <Toolname>` to run a local tool after https://github.com/dotnet/cli/pull/10980

**Comment by amis92** [comment]
Seeing as there is https://github.com/dotnet/announcements/issues/107, and the "documentation" is currently this issue, can we have some info about how to author these local tools? I'm quite interested in testing them in preview.

**Comment by wli3** [comment]
@amis92 authoring is the same as global tools which is available already https://docs.microsoft.com/en-us/dotnet/core/tools/global-tools-how-to-create

**Comment by poke** [comment]
Just out of curiosity: Where are local tools installed to? Are they installed into a local folder relative to the `dotnet-tools.json`? Or are they still installed into a global location so projects that depend on the same version of the same tool version can reuse the binary? What is the uninstall story in that case? Is there a way to list all local tool installations and uninstall them properly?

**Comment by wli3** [comment]
added "Update local tool"

**Comment by wli3** [comment]
Note the following are current implementation detail, that is subject to change. 

> Where are local tools installed to?

The asset are installed to [nuget global package folder](https://docs.microsoft.com/en-us/nuget/consume-packages/managing-the-global-packages-and-cache-folders).

> uninstall story

It will be removed from the manifest file. No change in nuget global package folder since it is maintained by a different set of command(detail in previous link).

> Is there a way to list all local tool installations and uninstall them properly?

If you mean "install" by downloading asset. There is no easy way to do so, however, you could browse your nuget global package folder if there is no override of that folder. At the same time, you could clean that folder using nuget command. However, that will also clean your other nuget packages like newtonsoft.json. (and cause a longer review next time a project uses it)

**Comment by tebeco** [comment]
@wli3
Update is a great idea and would be really effective with a full update option :

`dotnet tool update --local --all`

asking developper to manually list + update tool one by one is kinda harsh

I'm super lazy and if would add a nightly build on my CI just for that one

so every night the tooling is updated

**Comment by restrellasosa** [comment]
[
![Capture](https://user-images.githubusercontent.com/6253560/57991475-38906a80-7a7d-11e9-9ff6-5bad861fe8b6.JPG)
](url)

**Comment by restrellasosa** [comment]
How i can fix this. I try all! I dont know why and if anybody have knowlegde pls!

**Comment by tebeco** [comment]
@restrellasosa 
is that related to local tool ? or should it be in a new issue ?

**Comment by restrellasosa** [comment]
Local tools vs2019 v16.1
[cid:image002.png@01D50E9E.F9466720]

[cid:image006.png@01D50E9E.F9466720]

Enviado desde Correo<https://go.microsoft.com/fwlink/?LinkId=550986> para Windows 10

________________________________
De: TeBeCo <notifications@github.com>
Enviado: Sunday, May 19, 2019 11:56:52 PM
Para: dotnet/cli
Cc: Rigoberto Estrella Sosa; Mention
Asunto: Re: [dotnet/cli] Local Tools Early Preview Documentation (#10288)


@restrellasosa<https://github.com/restrellasosa>
is that related to local tool ? or should it be in a new issue ?

**Comment by tebeco** [comment]
it's visual studio right ?
Visual Studio is your IDE
Local tool are the name given to a specific tooling ship via nuget to extend `dotnet.exe`

if your issue is related to Visual Studio and it installation
Can delete your comments here and create an issue in Visual Studio feedback hub ?
for that just click the "Feedback" button in visual studio ? It's in the top right corner next to the "Live Share" button

**Comment by wli3** [comment]
@tebeco there is an existing issue https://github.com/dotnet/cli/issues/10860 and I'd like to add this feature.  However, this is unlikely to make in 3.0.100 release due to this is not a trivial feature.

**Comment by wli3** [comment]
@tebeco and @restrellasosa I moved the other issue to https://github.com/dotnet/cli/issues/11386 

I hope this thread only focus on bigger design of local tools

**Comment by hknielsen** [comment]
I like that its part of the dotnet tool, and local tools. But why not part of the csproj? We got away with the nuget file, but now adding a new file for this feels weird? 🙂

**Comment by jnm2** [comment]
@hknielsen To me it's a relief that it's not part of a csproj. I make use of this in a PowerShell build script where there is no one csproj that is more particularly connected to the build script than any other csproj.

**Comment by tebeco** [comment]
@hknielsen 
`dotnet tool` should not have any relation with `csproj`.
`dotnet tool` are CLI adding it to a csproj would force you to restore all the dependency tree for your business code.  
This means :
* the restore time will blow up
* that transitive of the CLI will impact your code (or fail the restore)
* transitive of your code can make the resolution fail

there are scenarios where you want to use a CLI on a repo without restoring the nuget from any

there are also scenario where you need the tooling in order to be able to restore.
this is how `fake` and `paket` works for `fsharp` and this is pretty neat

**Comment by JeremyMorton** [comment]
Does this hook into `dotnet restore` or `dotnet build`, at all?

We have a DotnetCliTool that is a code generator that we run before compiling C#, `dotnet restore` gets the package and the DotnetCliTool, so someone building a project that uses the generator just needs to run `dotnet build` and it gets what is needed.

If I understand this, switching to a local tool means that every developer *needs* to run `dotnet tool restore` every time they don't have the version of our generator that's checked into the .config\dotnet-tools.json file in our repo. They can't just have build do the right thing, correct?

**Comment by wli3** [comment]
If it is a generator that is part of and always part of the build, writing a Task and deliver it via nuget would be more appropriate.

Local tools operate on a directory level while "dotnet restore" and "dotnet build" operate on project level. Although you could add "dotnet tool restore" as part of your build by configuring your csproj.

**Comment by JeremyMorton** [comment]
We also have a Task in an assembly in a NuGet package for some build number logic. The big drawback of that is that the dll is held open by VS, so restoring when the project is opened in the IDE fails because it can't overwrite the assembly with a new version.

We get around this by having an init1.ps in the package (which only runs nowadays if you manually open the Package Manager Console in VS ☹️) to rename the in-use assembly. This is obviously a bit ugly, and only works if someone manually opens the Package Manager Console in VS.

```PowerShell
# Copy task binary file
$taskBinaryFileName = "CreateLocalPackages.dll"
$oldTaskBinaryPath = $taskBinaryFileName + ".old"
$olderTaskBinaryPath = $taskBinaryFileName + ".older"
$destTaskBinaryPath = Join-Paths $solutionDir.Directory $taskBinaryFileName
$packageTaskBinaryPath = Join-Paths $installPath "lib" "netstandard2.0" $taskBinaryFileName

if (Test-Path $olderTaskBinaryPath)
{
  Write-Host "Deleting $olderTaskBinaryPath"
  Remove-Item $olderTaskBinaryPath -ErrorAction:SilentlyContinue
  if (Test-Path $olderTaskBinaryPath)
  {
    Write-Host "Can't delete $olderTaskBinaryPath as it is in-use"
    Write-Host "Deleting $oldTaskBinaryPath"
    Remove-Item $oldTaskBinaryPath -ErrorAction:SilentlyContinue
  }
  else
  {
    Write-Host "Successfully deleted $olderTaskBinaryPath"
    if (Test-Path $oldTaskBinaryPath)
    {
      Write-Host "Moving $oldTaskBinaryPath to $olderTaskBinaryPath"
      Move-Item $oldTaskBinaryPath -Destination $olderTaskBinaryPath
    }
  }
}
else
{
  if (Test-Path $oldTaskBinaryPath)
  {
    Write-Host "Moving $oldTaskBinaryPath to $olderTaskBinaryPath"
    Move-Item $oldTaskBinaryPath -Destination $olderTaskBinaryPath
  }
}

if (Test-Path $destTaskBinaryPath)
{
  Write-Host "Moving $destTaskBinaryPath to $oldTaskBinaryPath"
  Move-Item $destTaskBinaryPath -Destination $oldTaskBinaryPath
}

Write-Host "Copying $packageTaskBinaryPath to $destTaskBinaryPath"
Copy-Item $packageTaskBinaryPath -Destination $destTaskBinaryPath
```

**Comment by AndersMalmgren** [comment]
Hi, we are migrating our enterprise application from 2.2 to 3.0. I found this post telling us to move to local tools if we use dot net CLI

https://github.com/dotnet/announcements/issues/107

So I found my way here. This does not seem like an replacement for donetcli tool at all? With dotnet cli we made a nuget package. And could reference it from project like 

`    <DotNetCliToolReference Include="dotnet.cqsgen" Version="1.0.22" />`

And run it like

```
  <Target Name="BuildContracts" AfterTargets="Build">
    <Exec Command="dotnet cqsgen $(OutputPath)IC.Eko.Core.Contracts.dll cqs.contracts.ts IC.Eko.Core.Contracts.Commands.Command;IC.Eko.Core.Contracts.Queries.Query" />
  </Target>
```

When a developer pulled latest version of our project from git and build the tool would invoke and do its code generation magic. I dont see how this can replace that? 

Thanks

**Comment by wli3** [comment]
@AndersMalmgren as the previous comments stated. If the tool is only used during build, could you try to make it a msbuild task?

At the same time, you could restore a local tool and run it with the similar fashion.

**Comment by wli3** [comment]
We used local tools for our tests. Here is a sample. Although you can start from "dotnet tool restore" instead of "dotnet new manifest" https://github.com/dotnet/sdk/blob/062d9a036ea335833e472d3b3c17174ba62c5c19/src/Tests/Directory.Build.targets#L37

**Comment by AndersMalmgren** [comment]
@wli3 I wasnt aware you could restore msbuild tasks using nuget? 
I solved it like this for now, 

```xml
  <Target Name="BuildContracts" AfterTargets="Build">
    <Exec Command="dotnet tool restore" />
    <Exec Command="dotnet SignalR.EventAggregatorProxy.AspNetCore.GlobalTool $(OutputPath)SignalR.EventAggregatorProxy.Demo.AspNetCore.dll $(MSBuildProjectDirectory)\wwwroot\js\events.js SignalR.EventAggregatorProxy.Demo.AspNetCore.EventTypeFinder" /> 
  </Target>
``´
though it feels a bit like a step back from CLI tool that you are forced todo a restore even though its fast becasue of nuget cache

**Comment by wli3** [comment]
@AndersMalmgren we acknowledge this short coming. And we do want to remove this step once some nuget feature is ready

**Comment by AndersMalmgren** [comment]
@wli3 Good to hear. I have found a showstopper sadly. We have private feeds. These interfere with nuget tool restore. Even though the tool is published to nuget.org it tries to use the private feeds and fails

**Comment by wli3** [comment]
--ignore-failed-sources should solve it

**Comment by AndersMalmgren** [comment]
@wli3 Thanks, that works. But I cant help to think the old CLI solution was more straight forward :D

**Comment by NickDarvey** [comment]
> If the tool is only used during build, could you try to make it a msbuild task

@wli3, if I understand correctly, an msbuild task would be a fine replacement for simple things but as soon as you need to use a dependency the advice I've read is [_don't build an msbuild task_](https://natemcmaster.com/blog/2017/11/11/msbuild-task-with-dependencies/#just-dont).

I don't quite understand the separation that's trying to be achieved here, (which may just be because I haven't been in the space long enough) but it seems to me using local tools as part of a build would be a common use case that should be designed for (with something like `DotNetCliToolReference`) rather than [requiring a workaround](https://github.com/dotnet/cli/issues/10288#issuecomment-545576270).

**Comment by stijnherreman** [comment]
Is this feature still in preview? Or is it the documentation that's in preview?

**Comment by wli3** [comment]
@stijnherreman only the documentation is in preview. It is shipped in 3.0.100 SDK.

We are working to convert this documentation to an official one
